src/frequenz/sdk/model/_manager.py

```diff
@@ -0,0 +1,273 @@
+# License: MIT
+# Copyright © 2023 Frequenz Energy-as-a-Service GmbH
+
+"""Load, update, monitor and retrieve machine learning models."""
+
+import abc
+import asyncio
+import logging
+import os
+import pickle
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any, Generic, TypeVar
+
+from frequenz.channels.file_watcher import EventType, FileWatcher
+
+from frequenz.sdk._internal._asyncio import cancel_and_await
+
+_logger = logging.getLogger(__name__)
+
+T = TypeVar("T")
+"""The model type."""
+
+
+class ModelRepository(abc.ABC):
+    """Provide an interface for building the model repository configuration.
+
+    The ModelRepository is meant to be subclassed to provide specific
+    implementations based on the type of models being managed.
+    """
+
+    @abc.abstractmethod
+    def _get_allowed_keys(self) -> list[Any]:
+        """Get the list of allowed keys for the model repository.
+
+        Returns:
+            a list of allowed keys.  # noqa: DAR202
+
+        Raises:
+            NotImplementedError: if the method is not implemented in the subclass.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def _build_path(self, key: Any) -> str:
```

Comment:
I would use `pathlib.Path` instead of `str` for paths.

```suggestion
    def _build_path(self, key: T) -> pathlib.Path:
```